agent/api/ecsclient/client.go

```diff
@@ -494,7 +507,39 @@ func (client *APIECSClient) buildContainerStateChangePayload(change api.Containe
 		statechange.ExitCode = aws.Int64(exitCode)
 	}
 
+	networkBindings := getNetworkBindings(change, shouldExcludeIPv6PortBinding)
+	// we enforce a limit on the no. of network bindings for containers with at-least 1 port range requested.
+	// this limit is enforced by ECS, and we fail early and don't call SubmitContainerStateChange.
+	if change.Container.GetContainerHasPortRange() && len(networkBindings) > ecsMaxNetworkBindingsLength {
+		return nil, fmt.Errorf("no. of network bindings %v is more than the maximum supported no. %v, "+
+			"container: %s "+"task: %s", len(networkBindings), ecsMaxNetworkBindingsLength, change.ContainerName, change.TaskArn)
+	}
+	statechange.NetworkBindings = networkBindings
+
+	return statechange, nil
+}
+
+// ProtocolBindIP used to store protocol and bindIP information associated to a particular host port
+type ProtocolBindIP struct {
+	protocol string
+	bindIP   string
+}
+
+// getNetworkBindings returns the list of networkingBindings, sent to ECS as part of the container state change payload
+func getNetworkBindings(change api.ContainerStateChange, shouldExcludeIPv6PortBinding bool) []*ecs.NetworkBinding {
 	networkBindings := []*ecs.NetworkBinding{}
+	// we do not return any network bindings for awsvpc and host network mode tasks.
+	if change.Container.GetNetworkMode() == networkModeAWSVPC || change.Container.GetNetworkMode() == networkModeHost {
+		return networkBindings
+	}
+	// hostPortToProtocolBindIPMap is a map to store protocol and bindIP information associated to host ports
+	// that belong to a range. This is used in case when there are multiple protocol/bindIP combinations associated to a
+	// port binding. example: when both IPv4 and IPv6 bindIPs are populated by docker and shouldExcludeIPv6PortBinding is false.
+	hostPortToProtocolBindIPMap := map[int64][]ProtocolBindIP{}
+
+	containerPortSet := change.Container.GetContainerPortSet()
+	containerPortRangeMap := change.Container.GetContainerPortRangeMap()
+
 	for _, binding := range change.PortBindings {
```

Comment:
Do we know for sure that each of those `PortBindings` will only container singular container and host port, and not a range? (asking because we seem to have `portBinding.ContainerPortRange` in `task.dockerPortMap` method. Although it looks like those are different structs, just want to make sure...

Reply:
yes, they're different structs. Docker's port-binding struct here, only has singular ports. While the bindings struct in `task.dockerPortMap` is parsing from the ECS task definition.

Reply:
I see! Overall it's been a bit challenging for me to follow the model translations from task payload -> docker input -> docker output -> STSC payload 😅